### ISSUE TITLE
Support for tunneling on 443/tcp at the same time as 1194/udp

### ISSUE THREAD

**Issue by vielmetti** [kylemanna/docker-openvpn#85]
So I'm looking at this comment in the Dockerfile

 Internally uses port 1194/udp, remap using `docker run -p 443:1194/tcp`

and I guess I'm confused. Can you support clients who use 443/tcp with the server inside thinking that it's talking UDP? I suppose the only way to know is try, but it seems so unlikely.

**Comment by ypid** [comment]
You will have to make sure that OpenVPN listens on that port on either UDP or TCP. UDP is the default.

**Comment by kylemanna** [comment]
It means that internally OpenVPN always listens on port 1194/udp inside the container.  To use it on a different external port you should use docker's port mapping feature (`docker run -p nnnn/proto`) to move it to another port.

To switch to tcp, the tcp option should also be passed so that docker listens on the tcp port.

Using TCP is not recommended except in very special cases.

**Comment by vielmetti** [comment]
Thanks to both of you, that's helpful.

I realized that TCP is not recommended for performance reasons (and a bunch
of other reasons), but the particular case I know I have to deal with is an
outbound firewall that filters quite aggressively that I have not yet been
able to get a VPN through.

On Mon, Dec 7, 2015 at 3:12 PM, Kyle Manna <notifications@github.com> wrote:

> It means that internally OpenVPN always listens on port 1194/udp inside
> the container. To use it on a different external port you should use
> docker's port mapping feature (docker run -p nnnn/proto) to move it to
> another port.
>
> To switch to tcp, the tcp option should also be passed so that docker
> listens on the tcp port.
>
> Using TCP is not recommended except in very special cases.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kylemanna/docker-openvpn/issues/85#issuecomment-162646260>
> .
>



-- 
Edward Vielmetti +1 734 330 2465
edward.vielmetti@gmail.com

**Comment by kylemanna** [comment]
That's one of the few valid use cases for the TCP VPN.

**Comment by joaocc** [comment]
Not sure if this was the original intent of the question, but I assumed the question could also apply on how to have a container that listens on both 443/tcp and 1194/udp, for those of us that use mostly UDP but may need to use TCP in some unexpected situations (hotels, customers, ...).
Any hints on this?
Thx

**Comment by dave-burke** [comment]
I would also like a way to run a second instance of the container on 443/tcp so that I can use 1194/udp in any sane environment, but have 443/tcp available in case I find myself somewhere very restrictive. As far as I can tell it would require two openvpn conf files (and two corresponding ovpn_env.sh files), and a way to choose which to run.

I'm working on a patch that adds a new argument to `ovpn_genconfig` that allows the user to specify a config suffix, so `ovpn_genconfig -s 443tcp` would generate `openvpn-443tcp.conf` and `ovpn_env-443tcp.conf`. Then `ovpn_run 443tcp` would use those instead of the defaults.

However, at this point it's not clear how to do that cleanly. For example, how does ovpn_run distinguish between an alternative config suffix and arguments that are supposed to be passed through to openvpn at the end of the script?

**Comment by kylemanna** [comment]
@dave-burke one way to do this would be to:
1. create a data container
2. configure the vpn for the first setup (udp perhaps)
3. copy the config to the second file name and manually edit
4. run two containers one with `ovpn_run --config openvpn.conf` and the second with `ovpn_run --config openvpn.tcp.conf`

But there are always a few ways to do this...

Another option would be to pass all the arguments on the openvpn command line and skip the `--config` option all together.  I'd imagine that'd get messy.

I'm not sure if openvpn does any prioritization of config file vs command line.  If it does, it may be possible to use one config file and specify the second container to run with the same config file and parameters like `--proto` to override the config file.

Also the openvpn man page does mention "configuration files can be nested to a reasonable depth." which may be the most robust option.

**Comment by dave-burke** [comment]
The first idea is roughly what I had been using as a workaround. It does work, I just thought it'd be nice to stramline it since I figured it would be a common use case. Maybe it isn't.

I actually like the idea of overriding with the `--proto` flag better, though, and it does work! Here's an example:
```bash
docker run --volumes-from ovpn-data --detach --publish 443:1194/tcp --cap-add=NET_ADMIN --name ovpn-443tcp --workdir "/etc/openvpn" openvpn ovpn_run --config openvpn.conf --proto tcp
```
The downside is that you have to specify anything that was otherwise being done for you. In my case it was just the `--workdir` from the Dockerfile, but for other people it could be the use of `--crl-verify` or `--client-config-dir` that `ovpn_run` does for you automatically.

Those "problems" would go away if `ovpn_genconfig` and `ovpn_run` allow the user to specify an alternate config name. Here's a (quick and untested) [example](https://github.com/dave-burke/docker-openvpn/commit/3477886f64ff9ef2511ca956a460556337276e2d) of what that would look like. Notice that the way I updated `ovpn_run` it checks to see if the first arg matches an existing config file. If so, it uses it and shifts so other args still get passed through. If not, it works like it always has.

The only thing it doesn't solve is the need to specify `--workdir` because it gets overridden from the Dockerfile as soon as you pass a custom CMD to `docker run`. Is there a reason it's set in the Dockerfile instead of as an ARG in ovpn_run?

**Comment by kylemanna** [comment]
@dave-burke comments follow:
> Those "problems" would go away if ovpn_genconfig and ovpn_run allow the user to specify an alternate config name. Here's a (quick and untested) example of what that would look like. Notice that the way I updated ovpn_run it checks to see if the first arg matches an existing config file. If so, it uses it and shifts so other args still get passed through. If not, it works like it always has.

The implementation looks clean and makes sense.  My typical concern with changes like this is that it complicates the container for first time users and increases the chance of breaking things as others submit pull requests.  To combat this concern I've employed unit tests, can you add a new unit test file that will generate the current and new configs as well as verify that correct things end up in the right place and connect to a test server?

See `tests/basic.sh` for an example of starting a docker server and connecting a docker client to it.  Should be able to verify that both protos then work as expected.  Could potentially even do them in parallel I think.

With your current approach, a brief blurb in `docs/` and a reasonable unit test I think you have the best implementation.

> The only thing it doesn't solve is the need to specify --workdir because it gets overridden from the Dockerfile as soon as you pass a custom CMD to docker run. Is there a reason it's set in the Dockerfile instead of as an ARG in ovpn_run?

Perhaps the simplest way to address this is to `cd $OPENVPN` in `ovpn_run` and friends.  Alternatively abolishing relative paths are useful as well, but probably more work.

**Comment by dave-burke** [comment]
Alright, so I've given this some more thought, and I think I have a solution I like better. Basically it involves updating only `ovpn_run` so that the user's args are combined with the generated args. The simple way to do that is just to append the user args like [this](https://github.com/kylemanna/docker-openvpn/compare/master...dave-burke:append-args). That works for me, but it would prevent someone from specifying a truly separate config file--they could only add a second config that openvpn would combine with the default (I think. I didn't actually test specifying two config files). A more verbose, but also more robust solution is to update `ovpn_run` to add arguments if and only if they weren't already specified by the user. That looks like [this](https://github.com/kylemanna/docker-openvpn/compare/master...dave-burke:merge-args).

With both solutions, the user can specify either

    docker run --volumes-from ovpn-data --detach --restart=always --publish 443:1194/tcp --cap-add=NET_ADMIN --name ovpn-443tcp openvpn ovpn_run --proto tcp

or something like

    docker run --volumes-from ovpn-data --detach --restart=always --publish 443:1194/tcp --cap-add=NET_ADMIN --name ovpn-443tcp openvpn ovpn_run --config /etc/openvpn/openvpn-tcp.conf

Either change would be a breaking change if someone:
1. Wanted to specify all config on the cli (did not want to use *any* config file).
2. Had an `$OPENVPN/ccd` directory and wanted to *avoid* having it set as the `--client-config-dir`.
3. Had a `crl.pem` file they *didn't* want to use with `--crl-verify`.

Those all seem pretty unlikely to me, but I guess you can be the judge.

Also, I do like the idea of using `cd $OPENVPN` in `ovpn_run` so I used that in both solutions.

What do you think @kylemanna? If you like either of those solutions, I can write the tests/docs and create a pull request.

**Comment by kylemanna** [comment]
I like your thinking for the `merge-args` branch.  If you can put together the test and docs in a pull request, I'd appreciate it!

> Either change would be a breaking change if someone:
> 1. Wanted to specify all config on the cli (did not want to use any config file).
> 2. Had an $OPENVPN/ccd directory and wanted to avoid having it set as the --client-config-dir.
> 3. Had a crl.pem file they didn't want to use with --crl-verify.

I think this is a reasonable assumption.  If someone has modified the docker image so much, they should either fork it or be running `openvpn` directly instead of `ovpn_run`.

As for the `-o` arg, lets leave that out for now as `--proto` can be overriden.  Less is more in my book.

**Comment by kylemanna** [comment]
Closed with #132
